README.md
# Automated PHP, Xdebug, Composer installer

**The latest release can be downloaded by clicking the button:**

[![Download latest](https://img.shields.io/badge/Download%20Installer-32A852?style=for-the-badge&logoColor=white&logo=DocuSign)](https://github.com/totadavid95/PhpComposerInstaller/releases/latest/download/PhpComposerInstaller.zip)

[Kattints ide a MAGYAR nyelvű dokumentációért! (Click here for HUNGARIAN documentation.)](https://github.com/totadavid95/PhpComposerInstaller/blob/master/README_hu.md)

- Always **extract** the downloaded *.zip* file, then run the extracted `PhpComposerInstall.exe` executable. **DO NOT EXECUTE DIRECTLY FROM THE ARCHIVE!**
- The installer requires [.NET 4.7.2](https://dotnet.microsoft.com/en-us/download/dotnet-framework/net472) to run. (This requirement is fulfilled on Windows 10 and 11 systems by default.)

This documentation always applies to the latest release only, previous versions can be viewed in the GitHub history.

Table of contents:
- [Automated PHP, Xdebug, Composer installer](#automated-php-xdebug-composer-installer)
  - [Project description](#project-description)
  - [Command line switches](#command-line-switches)
  - [Frequent issues](#frequent-issues)
    - [Bug report](#bug-report)
  - [Manual installation (on Windows)](#manual-installation-on-windows)
    - [Installing PHP](#installing-php)
    - [Installing Visual C++ Redistributable](#installing-visual-c-redistributable)
    - [Installing Xdebug for PHP](#installing-xdebug-for-php)
    - [Install Composer](#install-composer)
    - [Adding the installation to the Path environment variable](#adding-the-installation-to-the-path-environment-variable)
  - [Testing the installed tools](#testing-the-installed-tools)
  - [Manual installation (on Linux)](#manual-installation-on-linux)
  - [Help and support](#help-and-support)
  - [Contributing](#contributing)
  - [Releasing a new version](#releasing-a-new-version)

## Project description
This installer was primarily created for the students of *Server-side web programming* course taught also by me at *Eötvös Loránd Univerity* in Hungary as a utility that provides a straightforward way to install and set up PHP, Xdebug and Composer under Windows.

The installer does not require administrative privileges since it only installs for the logged in Windows user, therefore it can be executed on the lab computers. This saves us a lot of wasted time and unnecessary headaches.

The installer downloads the latest releases of the aforementioned software, then installs and configures them. If an installation has already been completed, the existing installation is removed first, and a clean installation of the latest version is performed. This mechanism can also be used to fix a broken installation or to perform an update.

The installer does not require administrative privileges since it only installs for the logged in Windows user.

## Command line switches

To view the available command line switches, execute the installer with the `--help` switch:

```shell
PhpComposerInstaller.exe --help
```

## Frequent issues
If the installer or any of the installed software doesn't work properly, the culprit could be one of the following:
- Some antivirus software mistakenly classify one of the installation steps (usually Composer) as malware (also known as a false positive alarm). If you use such software, try suspending its protection and running the installer again. The installer is not "viral", the source code can be found here on GitHub.
- It is possible that some kind of "fluctuation" occurred on your network and the download was interrupted. Try running the installer again, possibly at a later time.
- Using non-English characters in your Windows **username** may lead to errors in some cases, typically when using Tinker. There is already [a video on fixing this issue](https://www.youtube.com/watch?v=_zL3Pxnidj0) (in Hungarian).

### Bug report
If your issue is still unresolved after reading the above, see the [Help and support](#help-and-support) section.

## Manual installation (on Windows)
The installer performs the following steps programmatically.

### Installing PHP
- The official PHP webpage contains the currently supported versions of PHP: [https://windows.php.net/download](https://windows.php.net/download) 
- Download the __x86 / x64 Non Thread Safe__ version of your choice (preferably the latest one). Use the x86 version on 32-bit operating systems and the x64 version for a 64-bit OS.
- Extract the downloaded *.zip* archive.
- PHP requires a configuration file named __php.ini__ which is not included in the archive by default. Create a duplicate of the included __php.ini-development__ (which serves as an example configuration) and rename it __php.ini__.
- Open __php.ini__ and perform the following changes:
  - Find the following row:  `;extension_dir = "ext"`, and remove the `;` (comment sign) from the beginning. This will make PHP look for its extensions in its own directory. (Since PHP was installed locally, it would not look for the extensions in the right place by default.)
  - Next we will need to enable some extensions which will be required to run Composer, Laravel, etc properly later. Uncomment the following lines, thereby enabling the given extensions:
    ```ini
    ;extension=curl
    ;extension=fileinfo
    ;extension=mbstring
    ;extension=openssl
    ;extension=pdo_mysql
    ;extension=pdo_sqlite
    ```
- Save the __php.ini__ file when you're done.
- Copy the contents of the `php` folder to:
  `%LOCALAPPDATA%\Programs\php` (create the `php` folder in `Programs` if it does not exist yet).

Hint: copy `%LOCALAPPDATA%\Programs` to the address bar of your file explorer.

### Installing Visual C++ Redistributable

The [installation requirements](https://www.php.net/manual/en/install.windows.requirements.php) specify the Visual C++ Redistributable is needed to run PHP on Windows systems. If PHP gives a "dll error" on first execution, download the appropriate (32 or 64 bit depending on your OS architecture) version of the Visual C++ runtime:
- [vc_redist.x86.exe](https://aka.ms/vs/16/release/vc_redist.x86.exe) (32 bit)
- [vc_redist.x64.exe](https://aka.ms/vs/16/release/vc_redist.x64.exe) (64 bit)

### Installing Xdebug for PHP
- Select the Xdebug version corresponding your installed PHP version from [https://xdebug.org/download](https://xdebug.org/download) from the __Windows binaries__ section.
- Download Xdebug as a file named __php_xdebug.dll__.
- Copy the downloaded __php_xdebug.dll__ file to the __ext__ directory of your PHP installation.
- Open __php.ini__. The ini file follows this structure:
  ```ini
  [section_name]
  property_name = property_value
  ```
  - First enable Xdebug as a Zend extension under the `[PHP]` section of the file:
    ```ini
    [PHP]
    ; Here you will find lots of settings and comment, but just 
    ; insert this at the botton of the section before the next [section] begins:
    zend_extension = xdebug
    ``` 
  - Then configure the settings related to Xdebug by inserting a new `[xdebug]` section at the very bottom of your __php.ini__ file:
    ```ini
    [xdebug]
    xdebug.cli_color = 1
    xdebug.client_host = localhost
    xdebug.client_port = 9003
    ```
    - You may find a detailed description of all Xdebug settings in the [official documentation](https://xdebug.org/docs/all_settings).
- Save the __php.ini__ file.
- If a PHP process was running on the machine bound to this PHP configuration, then the process needs to be restarted in order to load the new PHP configuration.

### Install Composer
- Download the latest Composer v2 release be clicking here: [https://getcomposer.org/download/latest-2.x/composer.phar](https://getcomposer.org/download/latest-2.x/composer.phar)
- Copy the downloaded `composer.phar` file to: `%LOCALAPPDATA%\Programs\composer` (create the `composer` folder in `Programs` if it does not exist yet).
- Since this is a PHP archive (PHAR - **PH**p **AR**chive), Windows is not able to execute it directly (like it executes php.exe). We need to create a command file which will make the OS execute this PHAR in the PHP interpreter when opened. This will make the `composer` command available in the command line, so calling Composer will basically execute this `bat` file. Create a `composer.bat` file the __composer__ directory which contains the following one line:
  ```bat
  @php "%~dp0composer.phar" %*
  ```
- This will make the `composer.bat` call PHP to run Composer (which is a PHP archive) forwarding the parameters as well (hence %).

### Adding the installation to the Path environment variable
At this point the installation is almost done, but currently you can only call PHP and Composer by knowing their absolute path.

To make `php` and `composer` simply callable from any open command line, we need to add the **two folders** created inside `%localappdata/Programs` to the Path environment variable.

This is most easily done by typing "Edit environment variables for your account" into the Search bar. Alternatively you can press Win+R and execute `rundll32 sysdm.cpl,EditEnvironmentVariables`.

It is important to note that you only need to work in the upper part of the dialog, since those are the environment variable of the __current user__, and you are not able to modify the system-level properties on the lab machines anyway.

**You'll need to close any previously opened command prompts for the changes to take effect.**

## Testing the installed tools
You can test whether your installations were successful by executing `php -v` and `composer -V`. In both cases you should be able to see the version number of the installed product, and the PHP version command should also display some information about Xdebug.

You may get some more details about your PHP installation by copying the following code into a PHP file (e.g *text.php*):
```php
<?php
    // display all information about the PHP installation
    phpinfo();

    // display all information about the Xdebug installation
    xdebug_info();
?>
```

Now open a command line in the directory of the created `test.php` file (or use `cd` to get there and) and execute:  
`php -S localhost:3000 test.php`

You may now view [http://localhost:3000/](http://localhost:3000/) in a browser. This will display a through report about your PHP environment (including version number and the enabled extensions). Don't forget to close the command line. This will also stop the development server, of course.

There two tests can also be executed using the tools included in the `InstallTest` folder of the installer.

## Manual installation (on Linux)
Our automatic installer only works under Windows systems, but this guide may help you perform the same on Linux operating systems. The process is based on [this tutorial](https://tecadmin.net/how-to-install-php-8-on-ubuntu-20-04/) and has been tested under Xubuntu 20.04.

1. Update the package repository:
   
   ```shell
   sudo apt update
   ```
2. Install a package which helps manage PPAs:
   
   ```shell
   sudo apt install software-properties-common -y 
   ```
3. Add a new [PPA package repository](https://launchpad.net/~ondrej/+archive/ubuntu/php) which manages PHP version:
   
   ```shell
   sudo add-apt-repository ppa:ondrej/php 
   ```
4. Now the system can find how to install PHP, so we can start by installing the PHP interpreter itself. The version number must be entered, but it is also possible to deviate from the version number shown here as an example (at the time of writing the description, 8.1 is the latest), it must be used consistently, and the same must be entered afterwards everywhere.
  
   ```shell
   sudo apt install php8.1
   ```
5. Install the needed extensions for PHP 8.1. These are basically the following, but you can install additional ones if you want:
   
   ```shell
   sudo apt install php8.1-cli php8.1-xml php8.1-curl php8.1-fileinfo php8.1-mbstring php8.1-sqlite3 php8.1-xdebug
   ```
6. Check if PHP works using this command:
   
   ```shell
   php -v
   ```
   - As a result, you should see an output similar to the one below (naturally not with these exact version numbers):
     ```text
     PHP 8.1.2 (cli) (built: Jan 24 2022 10:42:33) (NTS)
     Copyright (c) The PHP Group
     Zend Engine v4.1.2, Copyright (c) Zend Technologies

CHANGELOG.md
# PHP & Composer Installer Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][keepachangelog], and this project adheres to [Semantic Versioning][semver].

## [1.2.5] - 2023-12-16

### Added

- Ability to get the installer version with `--version` switch.

### Changed

- Reworked command line option handling.
- Toggleable Composer installation.
- Installing Visual C++ Redistributable is now enabled by default to prevent errors when running PHP.
- Do not allow running the installer when another instance is already running.
- Code refactoring.

### Fixed

- Always waits for key press after the installer is finished.

## [1.2.4] - 2023-03-01

### Added

- Add "Contributing" section in README.
- Add "Releasing new versions" section in README.
- Enable `mysqli` extension when installing PHP.

### Changed

- Code refactoring (remove Hungarian comments, add documentation comments, etc.).
- Update license.
- Changelog now uses [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format.

### Fixed

- Enable `zip` extension when installing PHP. This is required for Composer to work properly with PHP 8.2 and above.
  Reported by Erik Tűri.
- Rewrite Release CI, because the previous one used deprecated practices.
- Fix nits in README.

## [1.2.3] - 2022-09-08

### Changed

- Version selector now starts with the latest PHP.
- New download button in README.
- Installing Xdebug and VC Redist is now optional.
  - You can use the following switches for Xdebug/VC Redist installation: `--with-xdebug`, `--with-vc-redist`
  
### Fixed

- Xdebug download link detection fix.

## [1.2.2] - 2022-07-02

### Changed

- Translate installer to English.

## [1.2.1] - 2022-04-02

### Changed

- Improve testing tools.
- Improve descriptions.

## [1.2.0] - 2022-04-02

### Added

- PHP version selector. (Based on PHP's official website.)
- Detect incompatible softwares.
- Uninstall switch: `--uninstall`
- Enabling CURL in PHP config to improve Composer performance.
- Automatic Visual C++ Redistributable installation.
- Checksum-based verification after every file download.
- Testing tools (`InstallTest` folder).

### Fixed

- Xdebug config fix.

### Changed

- More detailed console output.
- Improve CI.
- Update README.

[1.2.5]: https://github.com/totadavid95/PhpComposerInstaller/compare/v1.2.4...v1.2.5
[1.2.4]: https://github.com/totadavid95/PhpComposerInstaller/compare/v1.2.3...v1.2.4
[1.2.3]: https://github.com/totadavid95/PhpComposerInstaller/compare/v1.2.2...v1.2.3
[1.2.2]: https://github.com/totadavid95/PhpComposerInstaller/compare/v1.2.1...v1.2.2
[1.2.1]: https://github.com/totadavid95/PhpComposerInstaller/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/totadavid95/PhpComposerInstaller/compare/v1.1.0...v1.2.0

[keepachangelog]: https://keepachangelog.com/en/1.0.0/
[semver]: https://semver.org/spec/v2.0.0.html


PhpComposerInstaller/Program.cs
// PHP, Xdebug, Composer installer
// Created by: David Tota
// Contact: totadavid95@inf.elte.hu
// GitHub: https://github.com/totadavid95/PhpComposerInstaller

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.IO.Compression;
using System.Linq;

namespace PhpComposerInstaller {
    /// <summary>
    /// Installer program root class.
    /// </summary>
    internal class Program {
        private static Dictionary<string, Dictionary<string, string>> phpReleases;
        private static string selectedPhpRelease;
        private static string defaultPhpLocation = PHP.GetDefaultPhpInstallationLocation();
        private static string xdebugVersion;
        private static string defaultComposerLocation = Composer.GetDefaultComposerInstallationLocation();
        private static bool phpAlreadyInstalled = false;
        private static bool composerAlreadyInstalled = false;

        /// <summary>
        /// The dictionary of command-line options with their default values and descriptions.
        /// </summary>
        private static readonly Dictionary<string, Option> options = new Dictionary<string, Option>()
        {
            { "uninstall",  new Option(value: false,    description: "Uninstall PHP and Composer from the local AppData directory") },
            { "cleanup",    new Option(value: true,     description: "Delete the temporary files after installation") },
            { "composer",   new Option(value: true,     description: "Install Composer") },
            { "xdebug",     new Option(value: false,    description: "Install Xdebug") },
            { "vc-redist",  new Option(value: true,     description: "Install Visual C++ Redistributable") }
        };

        /// <summary>
        /// The option handler instance.
        /// </summary>
        private static OptionHandler optionHandler;

        /// <summary>
        /// Uninstalls previously installed PHP and Composer, if they are installed with this tool.
        private static void HandleUninstall() {
            Console.WriteLine("Phase 1: Uninstalling");
            Console.WriteLine("---------");

            Installer.RemovePhpFromLocalAppData();
            Installer.RemovePhpFromPathIfNecessary();

            Installer.RemoveComposerFromLocalAppData();
            Installer.RemoveComposerFromPathIfNecessary();
        }

        /// <summary>
        /// Checks if PHP and Composer are already installed, and if they are, what version they are.
        /// </summary>
        private static void HandlePriorCheck() {
            // Detecting architecture
            Console.WriteLine("  * Detected architecture: " + (Environment.Is64BitOperatingSystem ? "64 bit" : "32 bit"));

            // Searching for PHP and Composer
            var phpLocations = OS.FindProgramLocations("php");
            var composerLocations = OS.FindProgramLocations("composer");

            // Check typical installation locations
            if (phpLocations.Contains(defaultPhpLocation) || File.Exists(defaultPhpLocation)) {
                Console.WriteLine("  * Currently installed PHP version: " + PHP.GetPhpVersionByLocation(defaultPhpLocation));
                phpAlreadyInstalled = true;
            }
            if (composerLocations.Contains(defaultComposerLocation) || File.Exists(defaultComposerLocation)) {
                Console.WriteLine("  * Currently installed Composer version: " + Composer.GetComposerVersionByLocation(defaultComposerLocation));
                composerAlreadyInstalled = true;
            }

            // Searching for other PHP and Composer installations, which are not installed by this installer.
            // These installations are only matters if they are in the PATH environment variable, because they
            // may cause conflicts with our installation.
            var foreignPhpLocations = OS.FindProgramLocations("php").Where(
                location => !location.Equals(defaultPhpLocation)
            ).ToList();

            var foreignComposerLocations = OS.FindProgramLocations("composer").Where(
                location => !location.Equals(defaultComposerLocation)
            ).ToList();

            if (foreignPhpLocations.Count > 0 || foreignComposerLocations.Count > 0) {
                Console.WriteLine("  * ATTENTION! The installer detected a PHP/Composer installation that is included in the PATH environment");
                Console.WriteLine("    variable, but it was not installed by this installer. This installer also adds the installed PHP to the");
                Console.WriteLine("    PATH, however if there are two PHPs in this environment variable at the same time, it leads to a conflict.");
                Console.WriteLine("    One possible solution to this problem may be to remove the paths listed below from the PATH variable, or");
                Console.WriteLine("    rename the php.exe file in them to php7.4.exe/php8.0.exe or anything other than php.exe.");
                Console.WriteLine(" ");
                if (foreignPhpLocations.Count > 0) {
                    Console.WriteLine("    Detected PHP installs:");
                    foreach (var phpLocation in foreignPhpLocations) {
                        Console.WriteLine("      - " + phpLocation + " (" + PHP.GetPhpVersionByLocation(phpLocation) + ")");
                    }
                    Console.WriteLine(" ");
                }
                if (foreignComposerLocations.Count > 0) {
                    Console.WriteLine("    Detected Composer installs:");
                    foreach (var composerLocation in foreignComposerLocations) {
                        Console.WriteLine("      - " + composerLocation + " (" + Composer.GetComposerVersionByLocation(composerLocation) + ")");
                    }
                    Console.WriteLine(" ");
                }
            } else {
                Console.WriteLine("  * No problems found.");
            }
        }

        /// <summary>
        /// Handles the PHP release selection.
        /// </summary>
        private static void HandlePhpReleaseSelection() {
            Console.Write("  * Fetching currently supported PHP releases from the official php.net site... ");
            phpReleases = PHP.GetPhpReleases();
            Console.WriteLine("OK.");

            Console.WriteLine("  * Currently, these supported PHP editions are available. Please choose which one");
            Console.WriteLine("    you want to install!");
            if (phpAlreadyInstalled || composerAlreadyInstalled) {
                Console.WriteLine("    ATTENTION! Immediately after the selection, the installer overwrites/updates");
                Console.WriteLine("    the currently installed PHP/Composer in the %LOCALAPPDATA%/Programs directory!");
            }

            var phpVersions = phpReleases.Keys.ToArray();
            Array.Reverse(phpVersions);

            selectedPhpRelease = Selector.ShowSelectorMenu(phpVersions);
            Console.WriteLine("  * Selected: " + selectedPhpRelease);
        }

        /// <summary>
        /// Handles the creation of the temporary directory.
        /// </summary>
        private static void HandleTempDirectory() {
            if (Directory.Exists(Constants.TempDirectory)) {
                OS.DeleteDirectory(Constants.TempDirectory);
                Console.WriteLine($"  * Temp directory ({Constants.TempDirectory}) deleted.");
            }

            Directory.CreateDirectory(Constants.TempDirectory);

            string composerDir = Path.Combine(Constants.TempDirectory, "composer");
            Directory.CreateDirectory(composerDir);
            Console.WriteLine($"  * Temp directory ({composerDir}) created.");
        }

        /// <summary>
        /// Handles the downloading of the PHP and Xdebug binaries.
        /// </summary>
        private static void HandleDownloads() {
            Download.DownloadAndCheckFile(
                "Downloading PHP " + selectedPhpRelease + "... ",
                new Uri(phpReleases[selectedPhpRelease]["downloadlink"]),
                phpReleases[selectedPhpRelease]["checksum"],
                Path.Combine(Constants.TempDirectory, "php.zip")
            );

            if (optionHandler.IsOptionEnabled("xdebug")) {
                var xdebug = Xdebug.GetLatestPackage(selectedPhpRelease, phpReleases[selectedPhpRelease]["builtwith"]);
                xdebugVersion = xdebug["version"];
                Download.DownloadAndCheckFile(
                    "Downloading Xdebug for PHP " + selectedPhpRelease + " " + phpReleases[selectedPhpRelease]["builtwith"].ToUpper() + " NTS... ",
                    new Uri(xdebug["downloadlink"]),
                    xdebug["checksum"],
                    Path.Combine(Constants.TempDirectory, "xdebug.dll")
                );

                Download.DownloadFile(
                    "Downloading Xdebug " + xdebug["version"] + " source code... ",
                    new Uri("https://github.com/xdebug/xdebug/archive/refs/tags/" + xdebug["version"] + ".zip"),
                    Path.Combine(Constants.TempDirectory, "xdebug_src.zip")
                );
            }

            if (optionHandler.IsOptionEnabled("vc-redist")) {
                // This is required for PHP, see https://www.php.net/manual/en/install.windows.requirements.php
                Download.DownloadFile(
                    "Downloading Visual C++ Redistributable " + (Environment.Is64BitOperatingSystem ? "x64" : "x86") + " installer... ",
                    new Uri(
                        Environment.Is64BitOperatingSystem
                            ? "https://aka.ms/vs/16/release/vc_redist.x64.exe"
                            : "https://aka.ms/vs/16/release/vc_redist.x86.exe"
                    ),
                    Path.Combine(Constants.TempDirectory, "vc_redist.exe")
                );
            }

            if (optionHandler.IsOptionEnabled("composer")) {
                Download.DownloadAndCheckFile(
                    "Downloading latest Composer 2.x... ",
                    new Uri(Composer.GetDownloadLinkForLatestVersion()),
                    Composer.GetChecksumForLatestVersion(),
                    Path.Combine(Constants.TempDirectory, "composer.phar")
                );
            }
        }

        /// <summary>
        /// Handles the configuration of the PHP and Xdebug binaries.
        /// </summary>
        private static void HandleConfiguration() {
            // PHP
            Console.Write("  * Extracting PHP program files... ");
            ZipFile.ExtractToDirectory(
                Path.Combine(Constants.TempDirectory, "php.zip"),
                Path.Combine(Constants.TempDirectory, "php")
            );
            Console.WriteLine("OK.");

            if (optionHandler.IsOptionEnabled("xdebug")) {
                Console.Write("  * Extracting Xdebug configuration... ");
                ZipFile.ExtractToDirectory(
                    Path.Combine(Constants.TempDirectory, "xdebug_src.zip"),
                    Path.Combine(Constants.TempDirectory, "xdebug_src")
                );
                File.Copy(
                    Path.Combine(Constants.TempDirectory, "xdebug_src", $"xdebug-{xdebugVersion}", "xdebug.ini"),
                    Path.Combine(Constants.TempDirectory, "xdebug.ini")
                );
                Console.WriteLine("OK.");

                Console.Write("  * Copy xdebug.dll to php/ext directory... ");
                File.Copy(
                    Path.Combine(Constants.TempDirectory, "xdebug.dll"),
                    Path.Combine(Constants.TempDirectory, "php", "ext", "php_xdebug.dll")
                );
                Console.WriteLine("OK.");
            }

            Console.Write("  * Create and configure php.ini... ");
            File.Copy(
                Path.Combine(Constants.TempDirectory, "php", "php.ini-development"),
                Path.Combine(Constants.TempDirectory, "php", "php.ini"),
                true
            );

            string phpIniContent = File.ReadAllText(Path.Combine(Constants.TempDirectory, "php", "php.ini"));
            phpIniContent = phpIniContent.Replace(";extension_dir = \"ext\"", "extension_dir = \"ext\"");
            phpIniContent = phpIniContent.Replace(";extension=curl", "extension=curl");
            phpIniContent = phpIniContent.Replace(";extension=fileinfo", "extension=fileinfo");
            phpIniContent = phpIniContent.Replace(";extension=mbstring", "extension=mbstring");
            phpIniContent = phpIniContent.Replace(";extension=openssl", "extension=openssl");
            phpIniContent = phpIniContent.Replace(";extension=pdo_mysql", "extension=pdo_mysql");
            phpIniContent = phpIniContent.Replace(";extension=pdo_sqlite", "extension=pdo_sqlite");

            // Some students uses MySQL instead of the recommended database manager.
            phpIniContent = phpIniContent.Replace(";extension=mysqli", "extension=mysqli");

            // Fix for PHP 8.2 - previously, the zip extension was enabled by default,
            // but from PHP 8.2 we need to enable it manually.
            phpIniContent = phpIniContent.Replace(";extension=zip", "extension=zip");

            if (optionHandler.IsOptionEnabled("xdebug")) {
                phpIniContent = phpIniContent.Replace(";extension=xsl", ";extension=xsl\nzend_extension=xdebug");

                // See https://xdebug.org/docs/all_settings
                string xdebugDefaultConfig = File.ReadAllText(Path.Combine(Constants.TempDirectory, "xdebug.ini"));
                xdebugDefaultConfig = xdebugDefaultConfig.Replace(" — do not modify by hand", "");
                xdebugDefaultConfig = xdebugDefaultConfig.Replace(";xdebug.cli_color = 0", "xdebug.cli_color = 1");
                xdebugDefaultConfig = xdebugDefaultConfig.Replace(";xdebug.client_host = localhost", "xdebug.client_host = localhost");
                xdebugDefaultConfig = xdebugDefaultConfig.Replace(";xdebug.client_port = 9003", "xdebug.client_port = 9003");
                phpIniContent += "\n[xdebug]\n" + xdebugDefaultConfig;
            }

            File.WriteAllText(Path.Combine(Constants.TempDirectory, "php", "php.ini"), phpIniContent);
            Console.WriteLine("OK.");

            // Composer
            if (optionHandler.IsOptionEnabled("composer")) {
                Console.Write("  * Copy composer.phar to the composer directory... ");
                File.Copy(
                    Path.Combine(Constants.TempDirectory, "composer.phar"),
                    Path.Combine(Constants.TempDirectory, "composer", "composer.phar")
                );
                Console.WriteLine("OK.");

                Console.Write("  * Create runnable .bat file for Composer... ");
                Composer.GenerateComposerBatchFile();
                Console.WriteLine("OK.");
            }
        }

        /// <summary>
        /// Handles the installation of the PHP and Composer binaries. If the user
        /// has enabled --with-vc-redist, the Visual C++ Redistributable will be
        /// installed as well.
        /// </summary>
        private static void HandleInstallation() {
            Console.WriteLine("  * Stop PHP processes that conflicts with this installer... ");
            PHP.KillRunningPhpProcessesByLocation(defaultPhpLocation);

            if (optionHandler.IsOptionEnabled("vc-redist")) {
                Console.Write("  * Run Visual C++ Redistributable Installer... ");
                var proc = new Process {
                    StartInfo = new ProcessStartInfo {
                        FileName = Path.Combine(Constants.TempDirectory, "vc_redist.exe"),
                        Arguments = "/install /passive /quiet /norestart /log vc_redist.log",
                        UseShellExecute = false,
                        RedirectStandardOutput = true,
                        CreateNoWindow = true
                    }
                };

                proc.Start();
                proc.WaitForExit();
                Console.WriteLine("OK.");
            }

            Installer.CopyPhpToLocalAppData();
            Installer.AddPhpToPathIfNecessary();

            if (optionHandler.IsOptionEnabled("composer")) {
                Installer.CopyComposerToLocalAppData();
                Installer.AddComposerToPathIfNecessary();
            }
        }

        /// <summary>
        /// Cleans up the temporary directory if the user didn't specify the --no-cleanup argument.
        /// </summary>
        private static void HandleCleanup() {
            if (optionHandler.IsOptionEnabled("cleanup")) {
                if (Directory.Exists(Constants.TempDirectory)) {
                    OS.DeleteDirectory(Constants.TempDirectory);
                    Console.WriteLine($"  * Temp directory ({Constants.TempDirectory}) deleted.");
                }
            } else {
                Console.WriteLine("  * Skipped.");
            }
        }

        /// <summary>
        /// Handles the installation process.
        /// </summary>
        private static void HandleInstall() {
            Console.WriteLine("Phase 1: Diagnostics");
            Console.WriteLine("---------");
            HandlePriorCheck();
            Console.WriteLine(" ");

            Console.WriteLine("Phase 2: PHP release selection");
            Console.WriteLine("---------");
            HandlePhpReleaseSelection();
            Console.WriteLine(" ");

            Console.WriteLine("Phase 3: Download tools from the Internet");
            Console.WriteLine("---------");
            HandleTempDirectory();
            HandleDownloads();
            Console.WriteLine(" ");

            Console.WriteLine("Phase 4: Configure downloaded tools");
            Console.WriteLine("---------");
            HandleConfiguration();
            Console.WriteLine(" ");

            Console.WriteLine("Phase 5: Installing programs");
            Console.WriteLine("---------");
            HandleInstallation();
            Console.WriteLine(" ");

            Console.WriteLine("Phase 6: Delete temporary files");
            Console.WriteLine("---------");
            HandleCleanup();
            Console.WriteLine(" ");

            Console.WriteLine("Phase 7: Test your installation manually!");
            Console.WriteLine("---------");
            Console.WriteLine("  1.) Open the \"version.bat\" file in the InstallTest directory, then make sure that");
            Console.WriteLine("      the version number of PHP and Composer is displayed on the console.");
            Console.WriteLine(" ");
            Console.WriteLine("  2.) Open the \"phpinfo.bat\" file in the InstallTest directory, which will start a PHP server");
            Console.WriteLine("      and then open it in MS Edge. If this works well, the installation is probably successful.");
            Console.WriteLine("      You can also visit this page if you want to check any PHP settings.");
            Console.WriteLine(" ");
        }

        /// <summary>
        /// Main method of the program.
        /// </summary>
        static void Main(string[] args) {
            try {
                // Do not allow to run the installer multiple times at the same time.
                if (Process.GetProcessesByName(Process.GetCurrentProcess().ProcessName).Length > 1) {
                    throw new Exception("Another instance of the installer is already running.");
                }

                // Initialize the option handler.
                optionHandler = new OptionHandler(options);
                optionHandler.HandleArgs(args);

                // If the user specified the --uninstall argument, we should uninstall the programs.
                // Otherwise, we should install them.
                if (optionHandler.IsOptionEnabled("uninstall")) {
                    HandleUninstall();
                } else {
                    HandleInstall();
                }
            } catch (Exception exception) {
                // If any exception occurs, we want to display the error message and then exit the program.
                Console.WriteLine("Installation tool encountered an error:");
                Console.WriteLine(exception.Message);
            } finally {
                // TODO: Do not wait for keypress if the program is started from the command line.
                // This is only necessary if the program is started from the file explorer.
                Console.WriteLine("Press any key to exit...");
                Console.ReadKey();
            }
        }
    }
}
